// @workspace/db/src/schema/cash_balances.ts
import {
  pgTable,
  serial,
  varchar,
  real,
  timestamp,
  uniqueIndex,
  index,
} from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";

export const cashBalancesTable = pgTable(
  "cash_balances",
  {
    id: serial("id").primaryKey(),

    account: varchar("account", { length: 100 }).notNull().default("default"),
    currency: varchar("currency", { length: 10 }).notNull().default("USD"),

    balance: real("balance").notNull().default(0),

    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    uniqueIndex("cash_balances_account_currency_idx").on(table.account, table.currency),
    index("cash_balances_currency_idx").on(table.currency),
  ],
);

export const insertCashBalanceSchema = createInsertSchema(cashBalancesTable).omit({
  id: true,
  updatedAt: true,
});

export type InsertCashBalance = z.infer<typeof insertCashBalanceSchema>;
export type CashBalanceRow = typeof cashBalancesTable.$inferSelect;
