// @workspace/db/src/schema/confirmations.ts
import {
  pgTable,
  serial,
  varchar,
  integer,
  jsonb,
  timestamp,
  uniqueIndex,
  index,
} from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { requestsTable } from "./requests";

// ---------------------------------------------------------------------------
// Confirmation statuses
// ---------------------------------------------------------------------------

export const CONFIRMATION_STATUSES = [
  "pending_confirmation",
  "confirmed",
  "rejected",
  "expired",
] as const;

export type ConfirmationStatus = (typeof CONFIRMATION_STATUSES)[number];

// ---------------------------------------------------------------------------
// Summary types embedded in the jsonb column
// ---------------------------------------------------------------------------

export interface ConfirmationSummaryItem {
  type: string;
  status: string;
  amount?: number;
  currency?: string;
  sku?: string;
  productName?: string;
  quantity?: number;
  customerEmail?: string;
  vendorName?: string;
}

export interface ConfirmationSummary {
  title: string;
  sourceSystem: string;
  sourceProvider?: string;
  sourceEventType: string;
  sourceId?: string;
  operationCount: number;
  items: ConfirmationSummaryItem[];
  totals: {
    amount: number;
    currency: string;
    itemCount: number;
  };
}

// ---------------------------------------------------------------------------
// Validation result types
// ---------------------------------------------------------------------------

export interface StockCheckItem {
  sku: string;
  productName?: string;
  required: number;
  available: number;
  pass: boolean;
}

export interface DuplicateCheckItem {
  sourceId: string;
  existingOperationId: number;
  existingBatchId?: string;
}

export interface TotalsCheckResult {
  salesTotal: number;
  paymentsTotal: number;
  delta: number;
  pass: boolean;
}

export interface ValidationResults {
  passed: boolean;
  hasWarnings: boolean;
  checks: {
    stock: { passed: boolean; items: StockCheckItem[] };
    duplicates: { passed: boolean; items: DuplicateCheckItem[] };
    totals: { passed: boolean; result: TotalsCheckResult | null };
  };
}

// ---------------------------------------------------------------------------
// Table definition
// ---------------------------------------------------------------------------

export const confirmationBatchesTable = pgTable(
  "confirmation_batches",
  {
    id: serial("id").primaryKey(),

    /** Human-readable batch identifier. */
    batchId: varchar("batch_id", { length: 100 }).notNull(),

    status: varchar("status", { length: 50 })
      .notNull()
      .default("pending_confirmation")
      .$type<ConfirmationStatus>(),

    /** Human-readable summary ticket for operator review. */
    summary: jsonb("summary")
      .$type<ConfirmationSummary>()
      .notNull(),

    /** Validation check results. */
    validationResults: jsonb("validation_results")
      .$type<ValidationResults>()
      .notNull(),

    /** IDs of operations in this batch. */
    operationIds: jsonb("operation_ids")
      .$type<number[]>()
      .notNull()
      .default([]),

    /** FK to the triage request that spawned this batch. */
    requestId: integer("request_id").references(() => requestsTable.id, {
      onDelete: "set null",
    }),

    /** Reason provided when confirming or rejecting (optional). */
    reviewNote: varchar("review_note", { length: 1000 }),

    confirmedAt: timestamp("confirmed_at", { withTimezone: true }),
    rejectedAt: timestamp("rejected_at", { withTimezone: true }),
    /** When this batch expires if not reviewed. */
    expiresAt: timestamp("expires_at", { withTimezone: true }),

    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    uniqueIndex("confirmation_batches_batch_id_idx").on(table.batchId),
    index("confirmation_batches_status_idx").on(table.status),
    index("confirmation_batches_created_at_idx").on(table.createdAt),
  ],
);

export const insertConfirmationBatchSchema = createInsertSchema(
  confirmationBatchesTable,
).omit({ id: true, createdAt: true });

export type InsertConfirmationBatch = z.infer<typeof insertConfirmationBatchSchema>;
export type ConfirmationBatchRow = typeof confirmationBatchesTable.$inferSelect;
