// @workspace/db/src/schema/queue.ts
import {
  pgTable,
  serial,
  integer,
  varchar,
  timestamp,
  index,
} from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { requestsTable } from "./requests";

export const QUEUE_STATUSES = ["queued", "processing", "done", "cancelled"] as const;
export type QueueStatus = (typeof QUEUE_STATUSES)[number];

export const queueItemsTable = pgTable(
  "queue_items",
  {
    id: serial("id").primaryKey(),
    requestId: integer("request_id")
      .notNull()
      .references(() => requestsTable.id, { onDelete: "cascade" }),
    department: varchar("department", { length: 64 }).notNull(),
    priority: varchar("priority", { length: 16 }).notNull(),
    status: varchar("status", { length: 24 })
      .notNull()
      .$type<QueueStatus>()
      .default("queued"),
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("queue_items_department_idx").on(table.department),
    index("queue_items_status_idx").on(table.status),
    index("queue_items_priority_idx").on(table.priority),
    index("queue_items_request_id_idx").on(table.requestId),
  ],
);

export const insertQueueItemSchema = createInsertSchema(queueItemsTable).omit({
  id: true,
  createdAt: true,
});

export type InsertQueueItem = z.infer<typeof insertQueueItemSchema>;
export type QueueItemRow = typeof queueItemsTable.$inferSelect;
