// @workspace/db/src/schema/requests.ts
import {
  pgTable,
  serial,
  text,
  varchar,
  integer,
  real,
  timestamp,
  jsonb,
  boolean,
  index,
} from "drizzle-orm/pg-core";

import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";

export const requestsTable = pgTable(
  "requests",
  {
    id: serial("id").primaryKey(),
    originalText: text("original_text").notNull(),
    detectedLanguage: varchar("detected_language", { length: 8 }).notNull(),
    lumetraCode: text("lumetra_code").notNull(),
    finalResponse: text("final_response").notNull(),
    routeUsed: varchar("route_used", { length: 32 }).notNull(),
    department: varchar("department", { length: 64 }).notNull(),
    departments: jsonb("departments").$type<string[]>().notNull(),
    priority: varchar("priority", { length: 16 }).notNull(),
    confidence: real("confidence").notNull(),
    source: varchar("source", { length: 32 }).notNull(),
    rootsHit: jsonb("roots_hit").$type<string[]>().notNull(),
    timeSavedMin: integer("time_saved_min").notNull().default(0),
    costSavedUsd: real("cost_saved_usd").notNull().default(0),
    latencyMs: integer("latency_ms").notNull().default(0),
    aiUsed: boolean("ai_used").notNull().default(false),
    aiModel: varchar("ai_model", { length: 64 }),
    aiReasoning: text("ai_reasoning"),
    firewallAction: varchar("firewall_action", { length: 24 })
      .notNull()
      .default("send"),
    firewallFlags: jsonb("firewall_flags")
      .$type<string[]>()
      .notNull()
      .default([]),
    matchedRules: jsonb("matched_rules")
      .$type<string[]>()
      .notNull()
      .default([]),
    timestamp: timestamp("timestamp", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("requests_timestamp_idx").on(table.timestamp),
    index("requests_department_idx").on(table.department),
    index("requests_priority_idx").on(table.priority),
  ],
);

export const insertRequestSchema = createInsertSchema(requestsTable).omit({
  id: true,
  timestamp: true,
});

export type InsertRequest = z.infer<typeof insertRequestSchema>;
export type RequestRow = typeof requestsTable.$inferSelect;
// rebuild trigger
