// @workspace/.ignored_db/src/schema/inventory.ts
import {
  pgTable,
  serial,
  varchar,
  real,
  integer,
  timestamp,
  uniqueIndex,
  index,
} from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";

export const inventoryItemsTable = pgTable(
  "inventory_items",
  {
    id: serial("id").primaryKey(),

    sku: varchar("sku", { length: 255 }).notNull(),
    productName: varchar("product_name", { length: 500 }),

    quantityOnHand: integer("quantity_on_hand").notNull().default(0),
    quantityReserved: integer("quantity_reserved").notNull().default(0),

    unitCost: real("unit_cost"),
    currency: varchar("currency", { length: 10 }).notNull().default("USD"),

    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    uniqueIndex("inventory_sku_idx").on(table.sku),
    index("inventory_product_name_idx").on(table.productName),
  ],
);

export const insertInventoryItemSchema = createInsertSchema(inventoryItemsTable).omit({
  id: true,
  updatedAt: true,
});

export type InsertInventoryItem = z.infer<typeof insertInventoryItemSchema>;
export type InventoryItemRow = typeof inventoryItemsTable.$inferSelect;
