// @workspace/.ignored_db/src/schema/operations.ts
import {
  pgTable,
  serial,
  integer,
  varchar,
  real,
  boolean,
  text,
  timestamp,
  jsonb,
  index,
} from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";
import { requestsTable } from "./requests";

// ---------------------------------------------------------------------------
// Operation types — what the system recognises
// ---------------------------------------------------------------------------

export const OPERATION_TYPES = [
  "sale",           // Customer purchase / fulfilled order
  "purchase",       // Company procurement / PO
  "stock_update",   // Inventory change (low, stockout, adjustment)
  "payment",        // Payment event (success, failure, overdue, pending)
  "refund",         // Refund processed or denied
  "dispute",        // Chargeback, fraud alert, customer dispute
  "support_ticket", // Helpdesk / CRM ticket
  "contract",       // Contract lifecycle (expiry, renewal)
  "compliance",     // Compliance violation or audit exception
  "device_alert",   // IoT sensor / device event
] as const;

export type OperationType = (typeof OPERATION_TYPES)[number];

// ---------------------------------------------------------------------------
// Operation statuses
// ---------------------------------------------------------------------------

export const OPERATION_STATUSES = [
  "completed",  // Action finished successfully
  "pending",    // Waiting for action / approval
  "failed",     // Transaction / process failed
  "cancelled",  // Cancelled by user or system
  "open",       // Ticket / dispute in progress
  "escalated",  // Ticket / case escalated
  "overdue",    // Payment / delivery overdue
  "alert",      // IoT threshold / inventory alert
  "violation",  // Compliance breach
  "denied",     // Refund / PO denied
  "rejected",   // PO rejected / action rejected
  "fraud",      // Fraud detected
] as const;

export type OperationStatus = (typeof OPERATION_STATUSES)[number];

// ---------------------------------------------------------------------------
// Table definition
// ---------------------------------------------------------------------------

export const operationsTable = pgTable(
  "operations",
  {
    id: serial("id").primaryKey(),

    // ── What happened ──────────────────────────────────────────────────────
    type: varchar("type", { length: 50 })
      .notNull()
      .$type<OperationType>(),

    status: varchar("status", { length: 50 })
      .notNull()
      .$type<OperationStatus>(),

    // ── Source traceability ────────────────────────────────────────────────
    /** Which integration system emitted the event. */
    sourceSystem: varchar("source_system", { length: 50 }),
    /** Specific vendor (stripe, zendesk, debezium, …). */
    sourceProvider: varchar("source_provider", { length: 100 }),
    /** Original event_type string (e.g. "charge.failed"). */
    sourceEventType: varchar("source_event_type", { length: 200 }),
    /** External record ID for deduplication. */
    sourceId: varchar("source_id", { length: 255 }),
    /** FK to the triage request that was created. */
    requestId: integer("request_id").references(() => requestsTable.id, {
      onDelete: "set null",
    }),

    // ── Financial ──────────────────────────────────────────────────────────
    amount: real("amount"),
    currency: varchar("currency", { length: 10 }).default("USD"),

    // ── Item / inventory ───────────────────────────────────────────────────
    sku: varchar("sku", { length: 255 }),
    productName: varchar("product_name", { length: 500 }),
    quantity: integer("quantity"),

    // ── Parties ────────────────────────────────────────────────────────────
    customerName: varchar("customer_name", { length: 500 }),
    customerEmail: varchar("customer_email", { length: 500 }),
    vendorName: varchar("vendor_name", { length: 500 }),

    // ── Raw payload for full audit trail ───────────────────────────────────
    data: jsonb("data")
      .$type<Record<string, unknown>>()
      .notNull()
      .default({}),

    // ── Confirmation tracking ──────────────────────────────────────────────
    /** The confirmation batch this operation belongs to (if confirmation is required). */
    confirmationBatchId: varchar("confirmation_batch_id", { length: 100 }),

    // ── Execution tracking ─────────────────────────────────────────────────
    /** Whether an execution handler ran for this operation. */
    executed: boolean("executed").notNull().default(false),
    /** Structured record of side-effects applied (inventory deltas, balance changes). */
    executionLog: jsonb("execution_log").$type<Record<string, unknown>>(),
    /** Error message if the execution handler threw. */
    executionError: text("execution_error"),

    // ── Timestamps ─────────────────────────────────────────────────────────
    /** When the event actually occurred in the source system. */
    occurredAt: timestamp("occurred_at", { withTimezone: true }),
    /** When this row was inserted. */
    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
  },
  (table) => [
    index("operations_type_idx").on(table.type),
    index("operations_status_idx").on(table.status),
    index("operations_source_system_idx").on(table.sourceSystem),
    index("operations_source_id_idx").on(table.sourceId),
    index("operations_request_id_idx").on(table.requestId),
    index("operations_created_at_idx").on(table.createdAt),
  ],
);

// ---------------------------------------------------------------------------
// Insert schema & types
// ---------------------------------------------------------------------------

export const insertOperationSchema = createInsertSchema(operationsTable).omit({
  id: true,
  createdAt: true,
});

export type InsertOperation = z.infer<typeof insertOperationSchema>;
export type OperationRow = typeof operationsTable.$inferSelect;
